zerver/lib/events.py

```diff
@@ -371,10 +379,28 @@ def apply_event(state: Dict[str, Any],
                 event['flags'],
             )
 
-        # Below, we handle maintaining first_message_id.
         if event['message']['type'] != "stream":
+            if 'raw_recent_conversations' in state:
+                # Handle maintaining the recent_conversations data structure.
+                my_recipient_id = Recipient.objects.get(type=Recipient.PERSONAL,
+                                                        type_id=user_profile.id).id
+                if event['message']['recipient_id'] == my_recipient_id:
+                    sender_id = event['message']["sender_id"]
+                    recipient_id = Recipient.objects.get(type=Recipient.PERSONAL,
+                                                         type_id=sender_id).id
+                else:
+                    recipient_id = event['message']['recipient_id']
```

Comment:
This `recipient_id` calculation could potentially be moved to a function, since it's semi-repeated below, in the delete block.